### PR TITLE
High resource usage and server unexpected shutdown webhook

### PR DESCRIPTION
**Pull request by lastpeony**
https://github.com/ant-media/Ant-Media-Server/issues/4739

### CODE REVIEW

File: src/main/java/io/antmedia/AntMediaApplicationAdapter.java [hunk 1]
```diff
@@ -356,8 +360,43 @@ public Result importVoDFolder(String vodFolderPath) {
 
 		return result;
 	}
+
+	private void sendUnexpectedShutdownHook() {
```
Comment (mekya) [inline src/main/java/io/antmedia/AntMediaApplicationAdapter.java:364]: I think these hooks should be in the server side. Not on the application side

File: src/main/java/io/antmedia/AntMediaApplicationAdapter.java [hunk 2]
```diff
@@ -356,8 +360,43 @@ public Result importVoDFolder(String vodFolderPath) {
...
+		final String listenerHookURL = appSettings.getListenerHookURL();
+		if (listenerHookURL == null || listenerHookURL.isEmpty()) {
+			return;
+		}
+		logger.info("Setting timer to call unexpected server shutdown hook.");
+		vertx.setTimer(10, e -> notifyHook(listenerHookURL, null, HOOK_UNEXPECTED_SERVER_SHUTDOWN, null, null,
```
Comment (mekya) [inline src/main/java/io/antmedia/AntMediaApplicationAdapter.java:370]: Send server's hostname, address or similar things to identify the server in this hook

**Review by mekya — CHANGES_REQUESTED** [review]
Thank you for your work @lastpeony .

I just have some comments. Please check and let me know if you have some comments

File: src/main/java/io/antmedia/AntMediaApplicationAdapter.java (hunk 1 above)
Comment (lastpeony) [reply src/main/java/io/antmedia/AntMediaApplicationAdapter.java:364]: i agree.  do you have a class on your mind to send this hook?

File: src/main/java/io/antmedia/statistic/StatsCollector.java [hunk 3]
```diff
@@ -785,9 +829,91 @@ public boolean enoughResource(){
...
+			httpPost.setHeader(HttpHeaders.CONTENT_TYPE, ContentType.APPLICATION_JSON.getMimeType());
+			httpPost.setEntity(new StringEntity(json.toString()));
+
+			try (CloseableHttpResponse httpResponse = httpClient.execute(httpPost)) {
+				logger.info("POST Response Status:: {}", httpResponse.getStatusLine().getStatusCode());
+
+				HttpEntity entity = httpResponse.getEntity();
+				if (entity != null) {
+					// read entity if it's available
+					BufferedReader reader = new BufferedReader(new InputStreamReader(entity.getContent()));
```
[the remark this answers is not in the document]
Comment (mekya) [reply src/main/java/io/antmedia/statistic/StatsCollector.java]: @sonatype-lift ignoreall

**Review by mekya — APPROVED** [review]

File: src/main/java/io/antmedia/AntMediaApplicationAdapter.java [hunk 4]
```diff
@@ -356,8 +360,43 @@ public Result importVoDFolder(String vodFolderPath) {
...
+			}
+		});
+	}
+
+	private void sendHighResourceUsageHook(){
+		final String listenerHookURL = appSettings.getListenerHookURL();
+		if (listenerHookURL == null || listenerHookURL.isEmpty()) {
+			return;
+		}
+		final Queue<IScope> scopes = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (mekya) [reply src/main/java/io/antmedia/AntMediaApplicationAdapter.java]: outdated

**Review by mekya — APPROVED** [review]

**Pull request merged**